tmpl/wiki/mirage2-to-mirage3.md

```diff
@@ -0,0 +1,1416 @@
+
+## Porting the ping example to Mirage 3
+
+To illustrate how to solve some more common problems, let's look at the `ping` example, which has the following `config.ml`:
+
+```ocaml
+open Mirage
+
+let main =
+  let packages = [ "tcpip" ] in
+  let libraries = [ "tcpip.arpv4"; "tcpip.ethif"; "tcpip.ipv4" ] in
+  foreign
+    ~libraries ~packages
+    "Unikernel.Main" (console @-> network @-> clock @-> time @-> job)
+
+let () =
+  register "ping" [ main $ default_console $ tap0 $ default_clock $ default_time ]
+```
+
+Trying to `mirage configure` this fails:
+
+```
+mirage-skeleton/ping$ mirage configure -t unix
+mirage: unknown option `-t'.
+Usage: mirage configure [OPTION]... 
+Try `mirage configure --help' or `mirage --help' for more information.
+error while executing ocamlbuild -use-ocamlfind -classic-display -tags
+                        bin_annot -quiet -X _build-ukvm -pkg mirage
+                        config.cmxs
++ ocamlfind ocamlc -c -bin-annot -package mirage -o config.cmo config.ml
+File "config.ml", line 7, characters 5-14:
+Error: The function applied to this argument has type
+         ?keys:Mirage.key list ->
+         ?deps:Mirage.abstract_impl list -> 'a Functoria.impl
+This argument cannot be applied with label ~libraries
+Command exited with code 2.
+Hint: Recursive traversal of subdirectories was not enabled for this build,
+  as the working directory does not look like an ocamlbuild project (no
+  '_tags' or 'myocamlbuild.ml' file). If you have modules in subdirectories,
+  you should add the option "-r" or create an empty '_tags' file.
+  
+  To enable recursive traversal for some subdirectories only, you can use the
+  following '_tags' file:
+  
+      true: -traverse
+      <dir1> or <dir2>: traverse
+```
+
+because the function `foreign` no longer takes a `libraries` argument.
+
+### Libraries and packages
+
+`libraries` and `packages` were two optional string list arguments in Mirageversions >= 2.7.x but < 3.x.  `libraries` was for `ocamlfind` libraries, and `packages` for OPAM packages; both string lists were simply fed `ocamlfind` and `opam` commands respectively.
+
+The [documentation for `foreign`](http://docs.mirage.io/mirage/Mirage/index.html#val-foreign) will be a useful reference for us.  In Mirage 3, `packages` is now a variable of type `package list`.  One can get a `package` by calling [Mirage.package](http://docs.mirage.io/mirage/Mirage/index.html#val-package), which has the following signature:
+
+```
+val package : ?build:bool -> ?sublibs:string list -> ?ocamlfind:string list -> ?min:string -> ?max:string -> string -> package
+```
+
+`~build` can be set to true to note that the package is required only at build time, and is not needed for a running unikernel; `sublibs` and `ocamlfind` let the caller specify when items that previously were listed in `libraries` are required.  `min` and `max` allow the caller to require specific version numbers for packages.
```

Comment:
You say "Either `sublibs` or `ocamlfind` can be provided" - do we get `Invalid_argument` if both are present?